ducktape/tests/result_store.py

```diff
@@ -0,0 +1,175 @@
+        "cls_name": result.test_context.cls_name,
+        "function_name": result.test_context.function_name,
+        "injected_args": result.test_context.injected_args,
+        "run_time_sec": result.run_time,
+        "services": result.test_context.services,
+        "status": "pass" if result.success else "fail",
+        "error_msg": result.summary,
+        "data": result.data
+    }
+
+
+def create_session_datum(session_datum):
+    return session_datum
+
+
+class TestKey(object):
+    """
+    A TestKey identifies a particular test with a particular set of arguments
+    """
+
+    @staticmethod
+    def from_test_context(test_context):
+        return TestKey(test_context.module_name,
+                                test_context.cls_name,
+                                test_context.function_name,
+                                test_context.injected_args)
+
+
+    @staticmethod
+    def from_datum(datum):
+        return TestKey(datum["module_name"],
+                                datum["cls_name"],
+                                datum["function_name"],
+                                datum["injected_args"])
+
+    def __init__(self, module_name, cls_name, function_name, injected_args={}):
+        self.module_name = module_name
+        self.cls_name = cls_name
+        self.function_name = function_name
+        self.injected_args = injected_args
+
+    def __repr__(self):
+        return "~~".join(
+            [self.module_name,
+             self.cls_name,
+             self.function_name,
+             json.dumps(self.injected_args, separators=(',', ':'))])
+
+
+class FileSystemResultStore(object):
+    DB_NAME = "test_data.db"
+
+    def __init__(self, root_results_dir):
+        """Assume key is a string representing a path relative to root result directory"""
+        self.root_results_dir = root_results_dir
+        self.db_file = os.path.join(self.root_results_dir, "test_data.db")
+        self._bootstrap_store()
+
+    def _bootstrap_store(self):
+        with sqlite3.connect(self.db_file) as connection:
+            cur = connection.cursor()
+            cur.execute("""create table if not exists test_data (
+              id integer primary key autoincrement,
+              timestamp integer not null,
+              test_id text not null,
+              session_id text not null,
+              data text not null
+            );""")
+
+            cur.execute("""create table if not exists session_data (
+              id integer primary key autoincrement,
+              timestamp integer not null,
+              session_id text not null,
+              data text not null
+            );""")
+
+
+    def _serialize(self, datum):
+        return json.dumps(datum, separators=(',', ':'))
+
+    def _deserialize(self, datum_bytes):
+        return json.loads(datum_bytes)
+
+    def put(self, session_id, test_id, datum):
+        """"""
+        with sqlite3.connect(self.db_file) as connection:
```

Comment:
Would it be better to get the connection once and then just do:

```
with self.connection as connection:
```

to take advantage of the context manager transaction rollback? In fact, `self.connection` could just cache the connection so the `sqlite3.connect` is implicit with the first access.